Initial draft proposal of the new standard data type set for UAVCAN v1.0

pavel-kirienko opened the pull request:
This is my vision of the new data type set design built in accordance with [our new approaches](https://forum.uavcan.org/t/stockholm-summit-recap/170). One will notice that this new set is completely devoid of any equipment-specific data types, which is fully intentional. The old approach that we experimented with in v0 turned out to be deeply flawed because it forces one to design their applications in a very particular way, rendering the protocol unusable for use cases that were not foreseen when the set of data types was originally designed. More background info on the subject is available in the forum thread linked above.

The new approach is to provide a set of very generic data types that do not make any assumptions about the nature of the target application. The resulting type system is far less efficient, but the upside is that it is much more flexible and repurposable. Rather than trying to cater to every sensible application out of the box, we are now encouraging equipment vendors to provide custom vendor-specific message definitions for their equipment when higher efficiency or advanced functionality is desired; whereas simpler applications can be handled by using the standard set alone. I expect that there will be equipment that can be interfaced with using both generic data types (in a universal but not-so-efficient way) and vendor-provided data types (in a very vendor-specific but very efficient way). Similar patterns can be found in other systems, e.g. CANopen or USB.

I think it might be sensible to encourage vendors to publish their vendor-specific data types under very permissive licenses permitting other vendors to re-use their data type definitions to reduce fragmentation of the ecosystem. We could also come up with a set of simple domain-specific definitions to be used as a middle ground between the generic standard set and vendor-specific custom messages, although that should belong in a completely separate namespace outside of `uavcan.*`.

I want everyone involved to evaluate the new data type set carefully and pay particular attention to the following parts:

`uavcan.node.GetPorts`, `uavcan.node.GetPortInfo`, `uavcan.node.GetPortStatistics` - these data types provide a simple yet very powerful API that allows one to reconstruct the computational graph, collect the main communication metrics (input/output/errors) in real time, and verify that all nodes use compatible data types for communication with each other. Eventually, when the new GUI tool is around, it will be able to build things like this, showing the traffic intensity and transport errors in real time:

![12](https://user-images.githubusercontent.com/3298404/46830950-aa4b5c00-cdaa-11e8-8e16-266c465c73e2.png)
![cv_graph](https://user-images.githubusercontent.com/3298404/46830952-ac151f80-cdaa-11e8-908c-c44fd1c08487.png)

(regarding the term "port", see https://forum.uavcan.org/t/a-generic-term-for-either-subject-or-service/182)

The dynamic node ID allocation message was moved to a different namespace (named `uavcan.pnp`, short for plug-and-play). Now we have two versions of it: one legacy for CAN 2.0, and a new one for CAN FD and eventually other transports. The key difference is that the new message allows for much, much simpler allocation logic because it doesn't require the server and the client to engage in complex stateful multi-stage exchanges.

A new namespace `uavcan.primitive` was added, which contains primitive data types (integers, reals, arrays of those, etc). I expect it will be handy for prototyping, although I don't want to encourage reliance on it because such basic types lack semantic information making them hard to build robust interfaces with. Also, these are needed for the next feature:

The parameter API is gone. Meet its successor: `uavcan.register`. This is a pair of very simple services that allow the caller to manipulate the configuration parameters on the remote node, request its read-only named values such as calibration data and whatever else defined by the vendor, and trigger actions by writing values into special-function registers. The same interface can also be used to implement direct reads and writes of the target node memory remotely over the CAN bus. We use this approach (although with a completely different protocol over USB/RS485) for high-speed data collection and configuration parameter management elsewhere and it works quite well.

The data type `uavcan.node.ExecuteCommand` replaces overly specific definitions like `RestartNode` and `BeginFirmwareUpdate`. It also allows vendors to define custom commands there. This move should simplify implementation on resource-constrained nodes as one can get by with just one server where three used to be needed.

Finally, the brand new namespace `uavcan.si`, which is the main innovation here. Just a bunch of physical quantities. I personally think that the concept is solid but its implementation could use some refinement. For example, the data types may require renaming, some of them look clumsy (e.g., `uavcan.si.length.WideVector3TS` could certainly use a better name). Most types use `float32` underneath, but some have to rely on wide integers or `float64` because of insufficient precision/resolution/range of `float32`; such inconsistency is undesirable. As I expect this to be the most critical part of the new specification, I'd like everyone to ask to subject these definitions to maximum scrutiny.

Paging @kjetilkjeka @thirtytwobits @EShamaev @antoinealb @mhkabir @uxvrob @dagar

pavel-kirienko commented:
The CI is broken because the current Python DSDL compiler doesn't understand our new data types yet.

pavel-kirienko commented:
@antoinealb check out this video here please: https://forum.zubax.com/uploads/default/original/1X/3f8b3b136d03e9035ef95ca7928f1ba4a4f746fa.webm (you can skip to 1:20). It demonstrates the user experience that can be built in a very generic way (no assumptions made about the application) using this approach. If you have any specific concerns about generality/UI please share them, perhaps the documentation should be clarified?

antoinealb commented:
yes for parameter enumeration I am quite confident this can be done. I am more wondering about the "save to persistent storage" functionality that was there before. I don't see anything comparable to that in your new protocol ?

pavel-kirienko commented:
@antoinealb added a new standard command to support manual persistence management. I decided to avoid defining the load command as the use cases for that are not yet perfectly clear; interested vendors should rely on vendor-specific commands for now.

pavel-kirienko commented:
How do we represent covariance information in general-purpose standard interfaces using the new data type model? Our options are, as I see it:

* Use `uavcan.primitive.ArrayOfReal16` or  `uavcan.primitive.ArrayOfReal32`. These lack semantic information and timestamps.
* Define timestamped primitives, e.g. `uavcan.primitive.ArrayOfReal16TS`.
* Any other options?

pavel-kirienko commented:
Some definitions have been moved back to v0. I have posted my answers to the concerns raised by @kjetilkjeka here: https://forum.uavcan.org/t/on-subject-service-id-ranges/195/2

kjetilkjeka commented on uavcan/file/405.GetInfo.1.0.uavcan:
```diff
@@ -0,0 +1,23 @@
+#
```
This is changed from the old DSDL. It should be in v0.1 initially. We can, however, make a priority to test and stabilize this definition.

The file stuff has high priority, but we should prioritize the protocol stuff more. Let's make everything in file v0.1 initially and then do some testing and stabilize relatively quick. 

I don't want the essential stuff in the node namespace to have to wait for the file definitions before we are able to merge the PR.

pavel-kirienko replied:
Will change the version to 0.

kjetilkjeka commented on README.md:
```diff
@@ -11,4 +10,142 @@ UAVCAN is an open lightweight protocol designed for reliable communication in ae
+0       | 0         | Not a valid ID
+1       | 32767     | Dynamic
+61440   | 62803     | Vendor-specific public static
+62804   | 65535     | Standard static
```
How does this work with anonymous transfers? They need low subject IDs to work. Maybe 0-15 should be reserved as well for high priority static protocol transfers?

pavel-kirienko replied:
This is due to another bit of hidden reasoning I didn't share with the world yet, sorry. I should have managed my edits more carefully to avoid unnecessary confusion, I was just trying to take shortcuts to save time.

My idea was to invert the subject ID used in anonymous message transfers, moving them from [0, 7] to [65528, 65535]. That doesn't require us to split the ID range.

kjetilkjeka replied:
I don't like the "invert by default" approach, it's surprising behavior. We should also allow application designers to design data types usable with anonymous transfers as well. This means that we should also have some usable messages in the low ID range. What about one of the following things:

 - Reserve some IDs in the low ID range. That will give us the additional benefit of having some high priority IDs available.
 - Add an `invert` bit to the anonymous transfers making both the low ID and high ID scale usable. bit23 looks like a nice enough candidate.

pavel-kirienko replied:
> Add an `invert` bit to the anonymous transfers making both the low ID and high ID scale usable. bit23 looks like a nice enough candidate.

I like this. It would essentially define which side of the range the short ID should cling to: zero or 65535. The bit 23 seems to be situated in the middle of a reserved field, so I would rather use either 21 or 25, otherwise it seems wasteful.

kjetilkjeka replied:
Great! We should move this discussion to somewhere related to specification. The bit 23 was rooted in that we might want to increase number of bit usable in anonymous transfers at a later point? Anyway, let's take this to a different issue/thread.

kjetilkjeka commented on uavcan/node/430.GetInfo.1.0.uavcan:
```diff
@@ -0,0 +1,69 @@
+# This is the same value that is periodically published by the node.
+# The objective of its inclusion here is to ensure that the node did not restart since the last time
+# this message was received from it, because a restart may invalidate the node information.
+Status.1 node_status
```
`Status.1 status`?

kjetilkjeka commented on uavcan/node/430.GetInfo.1.0.uavcan:
```diff
@@ -0,0 +1,69 @@
+# Allowed characters are: a-z (lowercase ASCII letters) 0-9 (decimal digits) . (dot) - (dash) _ (underscore).
+# Node name is a reversed Internet domain name (like Java packages), e.g. "com.manufacturer.project.product".
+void2
+uint8[<64] name
```
Isn't a longer name useful here? <128?

pavel-kirienko replied:
Unlikely. This is not an arbitrary name, but rather a brief Java-package-like name, which tend to be short. The response here is the largest data structure in the standard set, I don't want to grow it even further.

kjetilkjeka commented on uavcan/node/430.GetInfo.1.0.uavcan:
```diff
@@ -0,0 +1,69 @@
+#
```
This definition has significantly changed from the old DSDL. It should be in v0.1 initially. This should be our top priority to get stabilized.

pavel-kirienko replied:
Will change to 0.

kjetilkjeka commented on uavcan/node/431.GetPorts.1.0.uavcan:
```diff
@@ -0,0 +1,35 @@
+#
```
This definition is brand new and untested. It should be in v0.1 initially. This should be our top priority to get stabilized.

kjetilkjeka commented on uavcan/node/432.GetPortInfo.1.0.uavcan:
```diff
@@ -0,0 +1,26 @@
+#
```
This definition has significantly changed from the old DSDL. It should be in v0.1 initially. This should be our top priority to get stabilized.

kjetilkjeka commented on uavcan/node/432.GetPortInfo.1.0.uavcan:
```diff
@@ -0,0 +1,26 @@
+
+# Flags describing the current port.
+# Note that a message port can be both subscriber and publisher at the same time.
+bool is_subscriber      # Message input port
```
We should really create different services for subject ID and service ID instead. This is a bit messy

pavel-kirienko replied:
We don't want to add more standard services than we really have to because they are taxing on the memory footprint -- you have to maintain a lot of state information to manage service requests. Seeing as this service serves only inspection needs and does not in any way help the node fulfill its utility function directly, node designers will be too keen to drop these if they find them too resource consuming or too difficult to implement.

kjetilkjeka replied:
How would two separate services be more taxing for ram footprint than one? Wouldn't one service mean that it would have the sum of calls that both services had if they were separate and ram use end up more or less the same?

pavel-kirienko replied:
See section 4.3.2 "State variables" of the current spec draft or [the old spec](https://uavcan.org/Specification/4._CAN_bus_transport_layer/#state-variables). We have a set of variables to maintain per every service we want to support, hence increased footprint.

kjetilkjeka commented on uavcan/node/62805.Status.1.0.uavcan:
```diff
@@ -0,0 +1,47 @@
+truncated uint3 mode
+
+# Optional, vendor-specific node status code, e.g. a fault code or a status bitmask.
+truncated uint19 vendor_specific_status_code
```
Is having a 19 bit status code weird? Should we instead just reserve 3 bits and have a 16 bit vendor specific status code?

pavel-kirienko replied:
My experience with the old vendor-specific status code definition used to often make me wish for more bits. Consider this an improvement based on empirical feedback.

kjetilkjeka commented on uavcan/node/PortID.1.0.uavcan:
```diff
@@ -0,0 +1,8 @@
+#
```
Not sure how useful the port abstraction is? What about having seperate services for subject and service?

kjetilkjeka commented on uavcan/node/IOStatistics.1.0.uavcan:
```diff
@@ -0,0 +1,8 @@
+#
```
This is not a high priority message to stabilize. Let's make it v0.1 or v0.0 to lower the work burden of initial stabilization?

thirtytwobits replied:
Agree. Demote to 0 for now.

kjetilkjeka commented on uavcan/node/SemanticVersion.1.0.uavcan:
```diff
@@ -0,0 +1,7 @@
+#
```
If this is going to be used on software as well patch version is relevant. What about:

```
uint6 major
uint5 minor
uint5 patch
```

pavel-kirienko replied:
We use semantic version tuples for the protocol itself, hardware, software, and data type definitions. The patch version can be sensibly used only for software. Do we really need to add a special case for that, given that there are other means of identifying a particular build of software? Namely, the version control system revision ID and the image hash.

kjetilkjeka replied:
Well, if we're going to call it SemVer we should have the ability to use a patch field. If we're going to use it for software it would be nice to have the patch field as well. I wouldn't even call it a "special case".

The alternative definition still uses 2 bytes and I think we can be quite comfortable with the ranges. I don't think we're worse off in any way with this alternative definition.

We can consider switching minor to the 6-bit field.

pavel-kirienko replied:
What I don't like here is that we would have an unused patch number in the other three use cases where it doesn't serve much purpose. SemVer was created for software, not vehicular systems, so maybe it's just not the best option here. How about any of the following options:

* We stop calling it `SemanticVersion` and call it just `MajorMinorVersion` or something else as long as it's generic. In the data type documentation we can explain how it should be used. References to semver.org are optional.
* We define another, abridged version struct alongside `SemanticVersion`, and use it for everything except software version. The software version will be represented by `SemanticVersion` with the patch number. The downside is that it's a new data type.

I am still not sure if the patch number is required because release management of embedded software is usually different compared to computer software: updates are always risky and expensive. I don't think the patch number can be sensibly applied to safety-critical systems, which are usually not updated unless there are major issues.

pavel-kirienko replied:
> > we would have an unused patch number

> Not really unused, it just happens to always be 0.

Which means that it is unused.

> A lot of security patches (major issues) doesn't change the external interface in any way.

It's absolutely irrelevant. In high safety assurance applications the very act of updating the firmware, even to the very same version down to the same binary image, can be a noticeable risk, significant enough to avoid that unless absolutely necessary. You don't even have to go that far, consider this: unless you are using a certified compiler or object code traceability, just a recompilation of the exact same codebase adds some amount of uncertainty and therefore risk.

> At the end of the day, I don't see much problem just putting a zero for the patch version.

The problem is that we're adding a new entity and not buying anything useful by doing that. With the existing API we can already do the following:

* Identifying the major and the minor versions of software, hardware, protocol, and each data type.
* Determining whether two deployed software images are of the exact same version via the image CRC. The patch version doesn't help here as it doesn't take into account the compilation process and ultimately depends on the developer's assessment of the impact of their changes, which can be unreliable.
* Mapping from the deployed software version to the corresponding revision of the source code via the VCS revision identifier.

Adding a new union for either major-minor or semver doesn't make sense because it adds more variance where it's unnecessary.

If a semantic version can't be used without the patch number, we can rename the type `SemanticVersion` to `MajorMinorVersion` (or similar) and keep everything else intact.

kjetilkjeka replied:
The rename route seems sensible, I'm convinced. As you suggested we can call it `MajorMinorVersion` or even just `Version` since this is how versions are used in DSDL. :+1:

kjetilkjeka commented on uavcan/pnp/65534.NodeIDAllocationData.1.0.uavcan:
```diff
@@ -0,0 +1,99 @@
+#
```
Why is it not possible to unite node ID allocation for CAN2.0 and other?

pavel-kirienko replied:
Anonymous multi-frame transfers are not possible, they can be only single-frame. The maximum payload length of a CAN 2.0 single-frame transfer is limited to 7 bytes. In order to allocate a node ID, we have to communicate the unique ID of the local node (which is 16 bytes large) to the allocator. Seeing as it won't fit in one CAN frame, the allocatee and the allocator have to engage in a very complex multi-stage exchange process, which also has certain [known hard-to-fix pitfalls](https://uavcan.org/Specification/6._Application_level_functions/#duties-of-the-leader).

The CAN FD-based protocol is vastly simpler: the allocatee can just send the request in one message, and the receive the response immediately.

kjetilkjeka replied:
That makes sense. Can we keep it in v0.1 just until the node allocator is tested and works with CAN-FD?

kjetilkjeka commented on uavcan/primitive/ArrayOfBits.1.0.uavcan:
```diff
@@ -0,0 +1,5 @@
+#
```
Can we make it so we don't need these silly definitions? Can we, for instance, instead allow primitive types (and arrays) to be sent on a subject ID as well as composite definitions?

pavel-kirienko replied:
I beg to differ. The definitions may look silly, but they are architecturally sound, unless I am mistaken somewhere. They explicitly define the set of basic types that can be exchanged over the bus, avoiding possible excessive fragmentation if one were to use oddly sized types like `uint51`. They operate within the framework of existing type system, not needing additional provisions from the specification, keeping it simple. They would be required anyway if we were to add timestamped primitives.

I don't see a compelling reason to add more complexity to the specification.

kjetilkjeka commented on uavcan/time/62804.Synchronization.1.0.uavcan:
```diff
@@ -0,0 +1,21 @@
+#
```
This is pretty much the same as for the old protocol. I've not tested these messages myself. Are they tested to the point that we're confident with publishing them as stabilized definitions? Or are there any weirdness we should discuss first?

pavel-kirienko replied:
I am not aware of any issues with these definitions. Maybe @jschall is?

pavel-kirienko commented on uavcan/diagnostic/65520.Text.1.0.uavcan:
```diff
@@ -0,0 +1,15 @@
+#
+
+# Standard severity, 2 bit wide.
+Severity.1 severity
```
I wasn't quite happy about LogLevel because of the implication that the text can be used only for logging purposes, whereas there is another valid use case related to real-time diagnostic displays. So I searched for a better name that would suit both use cases simultaneously without any bias towards either, and I ended up with Severity. The word `TRIVIAL` has the same origins (the other three were chosen to appease people with an aerospace background after we discussed that with Scott). I am open to better names, but I would prefer staying away from anything related to logging, if possible.

kjetilkjeka replied:
Are there any advantages to using the same data types for both these use cases?

pavel-kirienko replied:
I don't see any benefit in their separation. Logging and indication seem to be highly related.

kjetilkjeka replied:
With the current suggestion for this message, logging becomes more confusing. It's not trivial to know that `uavcan.diagnostics.Text` is a log message. And it becomes harder when you have to map some weird sounding (except for people with aerospace background) "Severities" to what you think of as log levels.

Perhaps if you elaborate on how this RT diagnostic display and how it relates to these severities it would help me see the whole picture? If you're diagnosing RT properties I think the `Trace` log level would be the intended way to do this. But then I guess the log message should have an optional timestamp as well?

pavel-kirienko replied:
I have removed this namespace from this PR to be discussed separately.

pavel-kirienko commented on uavcan/diagnostic/65520.Text.1.0.uavcan:
```diff
@@ -0,0 +1,15 @@
+
+# The actual useful text of the message.
+# Normally, messages should be kept as short as possible, especially those of high severity.
+uint8[<=200] text
```
We have to make our definitions compatible with resource-constrained nodes. That limits the maximum size of serialized messages -- I briefly mentioned that in the README, saying that the size should not be larger than some rather arbitrarily chosen limit. Now, messages and services differ in the amount of context that is maintained around them in the application when they are processed by the protocol stack. Simple services are cheap memory-wise, because usually at the time of responding to a service call the execution flow is situated somewhere in a callback from the protocol stack where the amount of contextual information is usually low, and as a result, the required stack depth is also low, which results in lower worst-case memory footprint. Certain other definitions, such as this message, for example, may be utilized in context-rich scenarios, e.g. if a log message is emitted from a sophisticated control loop, and so on, where the stack utilization can be high. To simplify application design and not force the designer to resort to various countermeasures, we need to account for that, and limit the worst case memory size of messages that are expected to affect worst-case stack utilization.

Other than that, there is no strict justification for 200 bytes exactly.

kjetilkjeka replied:
I don't think a lot of resource-constrained nodes will read diagnostics. For sending they can just use a lower bound internally.

On the other hand, the old definition was <= 90. Is there any reason to increase this number?

pavel-kirienko replied:
> On the other hand, the old definition was <= 90. Is there any reason to increase this number?

We don't have enough evidence to come up with a rigorously defined value here. I looked through some of the code I've written and decided that it might be useful to raise the limit, which I did. We can easily reduce it back or increase further (mind the above) if it is considered beneficial.

kjetilkjeka replied:
I don't like changing from one arbitrary value to a different arbitrary value. My proposal would be to either just use the whole uint8 for indexing and define it as [<=255]. This is the "we don't know so we just give as much leeway as possible" approach. The number will then be rooted in the fact that a longer dynamic array would need another byte for its length, and that it's very reasonable to think that 255 is long enough.

The alternative approach would delay stabilizing until we can gather data on how long these strings normally are and set a value based on this.

Since resource constrained nodes will not read log/diagnostics most of the time, the 255 bytes approach should be fine.

pavel-kirienko replied:
Increased to 255 max.

pavel-kirienko commented on uavcan/diagnostic/Severity.1.0.uavcan:
```diff
@@ -0,0 +1,9 @@
+# Generic message severity representation.
+#
+
+uint2 TRIVIAL  = 0      # No coloring (e.g. white/black)
```
Answered above, see Severity.

kjetilkjeka replied:
I think what you want to express in a logging situation is not "Severity" but rather a level. And to be useful in logging applications you will at least need to have something that corresponds to the following levels: Error, Warn, Info, Debug. And should probably have a Trace and Fatal as well for completeness.

The problem with this message used for logging is that Warning is the most severe "level" and there is no way to distinguish between Debug and Info (i guess both are trivial).

pavel-kirienko replied:
> I think what you want to express in a logging situation is not "Severity" but rather a level.

Correct me if I'm mistaken, but "log level" is a shorthand for "log severity level", so they are essentially the same thing. Perhaps we could use some input from a native speaker here.

> And to be useful in logging applications you will at least need to have something that corresponds to the following levels: Error, Warn, Info, Debug. And should probably have a Trace and Fatal as well for completeness.

The severity levels are defined to closely resemble those of the node status message, although this is admittedly not necessary. Those of the node status message, in turn, were chosen to appease people with the commercial aviation background. Paging @thirtytwobits 

Seeing as the shared bus is an expensive medium (unlike, say, log files), I don't think it will be as useful for messages of extremely low severity, such as Trace or Debug.

kjetilkjeka replied:
> Correct me if I'm mistaken, but "log level" is a shorthand for "log severity level", so they are essentially the same thing. Perhaps we could use some input from a native speaker here.

The problem is that when app devs want to map from their internal way of logging (using info, warning, error, etc) they will encounter an unnecessary step of converting between their log levels and the severity levels (and when they're seeking help in comments, all they get are some color names?). Some severities will also have to cater for multiple log levels. Both these things are unfortunate.

> Those of the node status message, in turn, were chosen to appease people with the commercial aviation background

I followed the discussion where we decided to use aviation inspired words for priority levels? When was this discussed in the context of node status? 

> Seeing as the shared bus is an expensive medium (unlike, say, log files), I don't think it will be as useful for messages of extremely low severity, such as Trace or Debug.

I don't think it's anything close to reasonable to have this level turned on by default. But it would be highly useful for bench testing (both debug and trace) and the ability to turn on debug for a misbehaving unit under some controlled tests are also useful.

pavel-kirienko commented:
I have temporarily disabled Travis here because we can't really fix it until the DSDL compiler is updated.

> The `si` namespace doesn't really look ergonomic either. Perhaps it is not quite ready to be stabilized directly from this PR?

I would really like to have all namespaces reviewed jointly, as some of them are interdependent (e.g. file and node, register and primitive). However, you are probably right in that the SI namespace is a rather special one as it is the major piece of the new design and I admitted above that it might need refinement. I am open to removing it from this PR, to re-introduce it again in the very next PR.

kjetilkjeka commented on README.md:
```diff
@@ -11,4 +10,142 @@ UAVCAN is an open lightweight protocol designed for reliable communication in ae
+
+From    | To        | Purpose
+--------|-----------|------------------------------------------------
+0       | 0         | Not a valid ID
```
What is the intended use case for an invalid ID?

pavel-kirienko replied:
Representing an unconfigured ID.

E.g., a node may have a configuration parameter selecting which Subject ID to use for a particular function. Setting it to zero means that the function is disabled.

kjetilkjeka replied:
It really sounds analogous to the mistake of `NULL`.

I think it's **much** more valuable to be certain that every `u16` is a valid subject ID. 

Having a `bool enabled` would be better code and less prone to error than having to check `subject_id != 0` everywhere.

This would be a DSDL convetion, meaning it would also be a breach of abstraction to use it in the core uavcan implementation.

pavel-kirienko replied:
Makes sense. Reverted this change back.

kjetilkjeka commented on uavcan/node/62805.Status.1.0.uavcan:
```diff
@@ -0,0 +1,47 @@
+
+# Abstract node health.
+uint2 HEALTH_OK         = 0         # Nominal. Color: none or green.
+uint2 HEALTH_ADVISORY   = 1         # Minor parameter deviation or minor issue. Color: blue/cyan.
```
The HEALTH_ADVISORY doesn't really make any intuitive sense. Sounds like a [committee](https://health.nt.gov.au/health-governance/health-advisory-committees)

pavel-kirienko replied:
This is derived from https://www.faa.gov/regulations_policies/rulemaking/committees/documents/media/TAEas-wcal-04232002.pdf

kjetilkjeka replied:
When I searched through the document I found a lot of references to "advisory light" which makes sense. I found little references to "node" or "health". Perhaps it's the "health" name that seems a bit off?

----
The first thing that comes to mind is the following

```python
# State used for indicating node status.
# Used in lights, panels, etc.
uint2 INDICATION_OK               = 0         # Nominal. Color: none or green.
uint2 INDICATION_ADVISORY   = 1         # Minor parameter deviation or minor issue. Color: blue/cyan.
uint2 INDICATION_CAUTION    = 2         # Major problem or failure. Color: amber/yellow.
uint2 INDICATION_WARNING   = 3         # Fault, loss of function. Color: red.
truncated uint2 indication
```

---

A different alternative would be to call the message `Heartbeat`, which is a pretty good description of what it is. And call the field `status`

```python
# The node status. Should be indicated with recommended color in lights, panels, etc.
uint2 STATUS_OK               = 0         # Nominal. Color: none or green.
uint2 STATUS_ADVISORY   = 1         # Minor parameter deviation or minor issue. Color: blue/cyan.
uint2 STATUS_CAUTION    = 2         # Major problem or failure. Color: amber/yellow.
uint2 STATUS_WARNING   = 3         # Fault, loss of function. Color: red.
truncated uint2 status
```

pavel-kirienko replied:
That makes sense. I have renamed the message to `Heartbeat` and renamed the field from `health` to `health_status`.

pavel-kirienko commented:
I accepted most of the feedback and removed the diagnostic namespace from this PR. I intend to re-introduce that in the very next PR.

> I followed the discussion where we decided to use aviation inspired words for priority levels? 

@kjetilkjeka It's weird, I recall Scott mentioned this somewhere on Github but now I can't find this, not even among my email notifications from Github. We also talked about that briefly at the Stockholm Summit on the first day before you got there.

@thirtytwobits it would really help if you vetted the new health codes we have here. It will also affect the new diagnostic message which I just removed from this PR: once we chose a particular way of representing node health, it would make sense to keep the diagnostic message within that framework too.

davids5 commented:
@pavel-kirienko - what are the implications for the bootloaders?

thirtytwobits commented on uavcan/pnp/65534.NodeIDAllocationData.0.1.uavcan:
```diff
@@ -0,0 +1,99 @@
+#     Rule B. On expiration of the Request Timer:
+#     1. Request Timer restarts with a random interval of Trequest (chosen anew).
+#     2. The allocatee broadcasts an Allocation request message, where the fields are populated as follows:
+#        node_id   - the preferred node ID, or zero if the allocatee doesn't have any specific preference
```
I'd like to see two "modes" for this process:

1. the currently described mode where the allocatee accepts any arbitrary node id provided.
2. A mode where the allocator requires that it recognize the provided unique_id. 

The second mode supports vehicle systems that must explicitly map all UIDs to node identifiers. While this may seem to defeat the purpose of the dynamic node allocation it is actually very useful where a given vehicle system is manufactured in large quantities. Imagine an eight wheeled rover where each wheel motor was the same hardware type but the controller had to map wheel motors 0-7 to node identifiers. The manufacturing line can't pre-assign node identifiers since any given motor may end up on any given wheel. Instead it makes sense for the vehicle assembly step to record the UID of each motor and provide the controller with a node-id to UID map. When the system is booted it provides each motor with a pre-determined node id these nodes come on-bus. If any motor is not able to obtain a node id then the map is wrong. It's better to fail to give the motor a node identifier then to give it the wrong identifier (i.e. the controller thinks it's controlling a wheel motor but a gimbal motor got the id instead. Hilarity ensues, cameras are destroyed).

To support this we could document an allocator mode where it always ignored "preferred id" and returned a failure code for "unrecognized UID".

pavel-kirienko replied:
Do such impostor nodes even need to know that they are being rejected? The allocator could just ignore them. The allocator will also be aware of the problem. The allocatee, if made aware, wouldn't be able to sensibly react to that error, would it?

thirtytwobits replied:
This is true but it might help a human trying to diagnose why a given wheel motor wasn't passing the "obtain a node id" state. It seems to provide additional diagnostic information without breaking any abstractions.

pavel-kirienko replied:
Can't the allocator just emit a log message to the bus? I really don't want to complicate this logic further.

thirtytwobits replied:
Let me ask a different question: how does the allocator know that a given allocation was accepted? For example: 
1. allocatee requests nodeid 1
2. allocator gets the request and queues an allocation message with the nodeid and the allocatee's UID
3. allocatee drops off the bus
4. allocator sends the queued message.

In this state the allocatee did not receive the node id allocation message but the allocator can't know this.

Again, I'm trying to preserve an important use of the current dynamic node allocation service where an allocator is recording the result of each operation to provide a central controller with the system to nodeid map. This new protocol doesn't seem robust enough to trust with this responsibility.

pavel-kirienko replied:
> how does the allocator know that a given allocation was accepted?

When the allocator does the actual allocation, it adds a new record to its allocation table (it works the same way for centralized and distributed allocators; in the latter case the act of "adding a new record" involves a bit of shared state synchronization, but the general idea holds and the allocatees are not aware of the difference). If the allocatee falls off the bus (sounds like a roadside accident), the allocator can't be aware of that, but eventually when the allocatee is back online it will request an allocation again. The allocator will see that there is an entry already in the allocation table, and serve the allocatee with the same node ID as before. The whole allocation protocol is entirely idempotent, which makes it robust against the kind of problem you're describing.

> Again, I'm trying to preserve an important use of the current dynamic node allocation service where an allocator is recording the result of each operation to provide a central controller with the system to nodeid map. This new protocol doesn't seem robust enough to trust with this responsibility.

The logic of the protocol is still the same. I just removed an unnecessary follow-up stage to make it simpler and more robust.

kjetilkjeka commented on README.md:
```diff
@@ -11,4 +10,128 @@ UAVCAN is an open lightweight protocol designed for reliable communication in ae
+--------|-----------|------------------------------------------------
+0       | 0         | Not a valid ID
+1       | 32767     | Dynamic
+61440   | 62803     | Vendor-specific public static
```
We discussed this further in https://forum.uavcan.org/t/on-subject-service-id-ranges/195

We should call it something like "Application Specific (Dynamic)" to not give the wrongful illusion that using it with static (application specific) IDs are not OK.

kjetilkjeka replied:
"The vendor specific public static" is fine. It's the "dynamic" I have a problem with (as it's not always dynamic). We should change that one to something like "Application specific".

kjetilkjeka commented on uavcan/node/62805.Heartbeat.1.0.uavcan:
```diff
@@ -0,0 +1,47 @@
+uint2 HEALTH_STATUS_ADVISORY = 1    # Minor parameter deviation or minor issue. Color: blue/cyan.
+uint2 HEALTH_STATUS_CAUTION  = 2    # Major problem or failure. Color: amber/yellow.
+uint2 HEALTH_STATUS_WARNING  = 3    # Fault, loss of function. Color: red.
+truncated uint2 health_status
```
I like this change a lot. But I feel there is still weirdness coming from the mix of "health" with the aviation words. When I read the documents I feel these words are intended to be grouped with "light", "staus" or "indicator". Is there any reason `health_status` was used over just `status` or `status_indication`

kjetilkjeka commented on uavcan/node/62805.Heartbeat.1.0.uavcan:
```diff
@@ -0,0 +1,47 @@
+truncated uint3 mode
+
+# Optional, vendor-specific node status code, e.g. a fault code or a status bitmask.
+truncated uint19 vendor_specific_status_code
```
I guess `node_specific_data` is a better name? As the same vendor might use it differently for different products?

Can we use one of the bits to make this field optional?

```python
truncated uint18[<=1] node_specific_data
```

Alternatively, If we wanted to leverage the CAN-FD capabilities without messing things up for CAN2.0B we could perhaps do the following:

NodeSpecificStatusCode:
```python
@union 
uint64[0] none # Is this a specification compliant way of creating a zero sized variant? We should probably allow `void0` or`()` to avoid this.
uint8[2] data_short # (alternatively `uint16 data_short`) The shorter alternative for vendor specific data.
uint8[8] data_long # (alternatively `uint64 data_long`) The longer alternative for vendor specific data. Highly recommended to avoid unless it's known that the underlying protocol will be able to send the whole frame in a single transfer.
# We could also have a `uint8[<=255] string` variant if we wanted to go yolo, of course urging users not to send strings that will cause a multi frame transfer.
```

---

I think this flexibility would be ok as this code/data is used for diagnosis/logging/debuging and not read by other nodes. A tool could for instance plot 64 bits and gray out 48 of them if the short variant was used. Or gray out all of them if the none variant was used.

pavel-kirienko replied:
Displaying this information is not the only major use case. The other one is other nodes checking the vendor-specific status code automatically, e.g. for extended application-specific self-diagnostics. Unions are flexible and fun, but they really add a lot of extra complexity and extend the testing matrix, so I would only use them when absolutely necessary. Vote for static typing.

![image](https://user-images.githubusercontent.com/3298404/47171260-e7ff3600-d310-11e8-985e-fe620eff2b63.png)

kjetilkjeka commented on uavcan/time/62804.Synchronization.1.0.uavcan:
```diff
@@ -0,0 +1,21 @@
+# Network-wide time synchronization message.
+# Any node that publishes timestamped data should use this time reference.
+#
+# This message should always be published at the highest priority level.
```
Perhaps we should reserve 16 Ports with the lowst ID and use one of them for this message?

pavel-kirienko replied:
I am not a big fan of splitting the ranges like that. The time synchronization protocol doesn't really require the highest priority, the comment is misleading and I added it a long time ago by mistake. Let me remove it now. The background is available at https://www.can-cia.org/fileadmin/resources/documents/proceedings/1994_gergeleit.pdf

pavel-kirienko commented on uavcan/node/62805.Heartbeat.1.0.uavcan:
```diff
@@ -0,0 +1,47 @@
+uint32 uptime                       # [second]
+
+# Abstract node health.
+uint2 HEALTH_STATUS_OK       = 0    # Nominal. Color: none or green.
```
Updated. I had to commit the change manually because the GitHub-generated diff is incorrect.

pavel-kirienko commented on README.md:
```diff
@@ -11,4 +10,141 @@ UAVCAN is an open lightweight protocol designed for reliable communication in ae
+The namespace `uavcan.si` contains a collection of generic data types describing commonly used
+physical quantities.
+
+Some of the messages are time stamped, in which case the time stamp is always situated at the end in order
```
In this section we're talking about data types. They don't know about CAN FD or anything about the transport layer in general, so when we say "at the end", we mean the last field of the data type.

pavel-kirienko commented on README.md:
```diff
@@ -11,4 +10,141 @@ UAVCAN is an open lightweight protocol designed for reliable communication in ae
+unless a different multiplier is explicitly specified in the definition (e.g., nanosecond).
+
+All coordinate systems are right-handed.
+In relation to body, the preferred standard is as follows: **X** -- forward, **Y** -- right, **Z** -- down.
```
We use a different coordinate frame: NED instead of ENU, because the formed is more widespread in aerospace. This was discussed at length years ago (on Skype, no records are available anymore), can't change easily.

thirtytwobits replied:
Okay. I like aerospace.

pavel-kirienko commented on README.md:
```diff
@@ -11,4 +10,141 @@ UAVCAN is an open lightweight protocol designed for reliable communication in ae
+Names of the time stamped messages have the suffux `TS`.
+
+All units follow the [International System of Units](https://en.wikipedia.org/wiki/International_System_of_Units).
+All units are unscaled basic units of measure (e.g., meters rather than kilometers, kilograms rather than milligrams),
```
We should be consistent above all. I strongly disagree with degrees Celsius and in fact I am going to sign a petition to the European Council to adopt Kelvin unionwide.

thirtytwobits replied:
(wonders if this is a joke?)

pavel-kirienko commented on uavcan/file/405.GetInfo.0.1.uavcan:
```diff
@@ -0,0 +1,23 @@
+
+Path.1 path
+
+---
```
*but my consistency*

Let's vote on this comment. Vote up to support Scott's proposition, vote down if you care about consistency as much as I do.

pavel-kirienko replied:
btw this works with the current spec:

```text
---#-------------------------------------
```

pavel-kirienko commented on uavcan/file/405.GetInfo.0.1.uavcan:
```diff
@@ -0,0 +1,23 @@
+Error.1 error
+
+# File size in bytes. Should be set to zero for directories.
+uint40 size
```
On a 64-bit platform it will be represented by a `uint64_t` anyway. We could make the field `truncated` though.

pavel-kirienko commented on uavcan/file/Error.1.0.uavcan:
```diff
@@ -0,0 +1,18 @@
+int16 NOT_FOUND         = 2
+int16 IO_ERROR          = 5
+int16 ACCESS_DENIED     = 13
+int16 IS_DIRECTORY      = 21 # I.e. attempted read/write on a path that points to a directory
```
`IS_DIRECTORY` is mapped from [`EISDIR`](https://docs.python.org/3/library/errno.html#errno.EISDIR). Do we really want to move away from that?

pavel-kirienko commented on uavcan/file/405.GetInfo.0.1.uavcan:
```diff
@@ -0,0 +1,23 @@
+uint40 size
+
+# If such entry does not exist, all flags should be cleared/ignored.
+bool is_file_not_directory  # True if file, false if directory
```
It used to be used in `ListDirectory`, being redundant. I have removed it from there, which left just one usage in `GetInfo`, so I decided to unroll it in-place for simplicity.

pavel-kirienko commented on uavcan/file/406.ListDirectory.0.1.uavcan:
```diff
@@ -0,0 +1,36 @@
+# index of the next request, the next returned item (or several, if more items were inserted) will repeat the ones
+# that were listed earlier. The caller should handle that properly, either by ignoring the repeated items or by
+# restarting the listing operation from the beginning (index 0).
+#
```
Too complex, doesn't provide much value: even if the snapshot is fully consistent within itself, by the time it is delivered to the client it might be obsolete.

pavel-kirienko commented on uavcan/file/409.Write.1.0.uavcan:
```diff
@@ -0,0 +1,26 @@
+
+Path.1 path
+
+uint8[<=128] data
```
1. It's a power of 2, which works well with bootloaders, which typically have to deal with power-of-two sized ROM sectors.
2. It's small enough to keep the total size of the encoded request under the limit of 364 bytes. The limit is based on the largest data structure used in the standard set (`node.GetInfo` response).
3. The length field is 8 bit wide, manual alignment not required.

pavel-kirienko commented on uavcan/file/Path.1.0.uavcan:
```diff
@@ -0,0 +1,10 @@
+
+uint8 SEPARATOR = '/'
+
+uint8[<=200] path
```
I was seriously considering to reduce it because 200 bytes is a lot of space and we're taking away from the payload in the `file.Write` service. With this change, `file.Write` would be almost 400 bytes large, quite taxing for small nodes. Do we really want that? Especially considering that:

* The length field will be 9 bit wide
* Who cares about Windows

pavel-kirienko commented on uavcan/file/Path.1.0.uavcan:
```diff
@@ -0,0 +1,10 @@
+# A file system path encoded in UTF8.
+# The only valid separator is the forward slash "/".
+# A single slash ("/") refers to the root directory.
+#
```
This is a topic of another lengthy debate. I suggest we just require real paths and add a note that the restrictions can be relaxed in the future.

pavel-kirienko commented on uavcan/file/407.Delete.1.0.uavcan:
```diff
@@ -0,0 +1,10 @@
+#
```
Adding a dedicated service for each is wasteful. There should be a generalized service for all generic file manipulation operations; for example `file.Move` or `file.Alter`, which would contain the path of the file, the output path (non-empty to move/copy (option flags galore), empty to delete, same to touch). Not sure about truncation, it seems to be doable with `file.Write`. Does it make sense? Are we better off postponing this feature, which means removing the `file` namespace from this PR?

thirtytwobits replied:
I think we either need to keep file almost exactly the same (i.e. just a namespace change) or we should spend more time getting it right. The value of the former is to ease porting to v1 whereas the latter could provide a more future-proof API that had more general utility.

If we want to modify it I'd remove from this PR and do it as a separate one.

pavel-kirienko commented on uavcan/file/Error.1.0.uavcan:
```diff
@@ -0,0 +1,18 @@
+int16 NOT_FOUND         = 2
+int16 IO_ERROR          = 5
+int16 ACCESS_DENIED     = 13
+int16 IS_DIRECTORY      = 21 # I.e. attempted read/write on a path that points to a directory
```
This is a duplicate of the above, marking resolved.

pavel-kirienko commented on uavcan/file/Error.1.0.uavcan:
```diff
@@ -0,0 +1,18 @@
+int16 ACCESS_DENIED     = 13
+int16 IS_DIRECTORY      = 21 # I.e. attempted read/write on a path that points to a directory
+int16 INVALID_VALUE     = 22 # E.g. file name is not valid for the target file system
+int16 FILE_TOO_LARGE    = 27
```
Again, this is mapped from Linux error codes. Do we want to keep following that or do we just define our own errors?

pavel-kirienko commented:
I would like to clarify that having all v1.0 data types merged doesn't really mean that they are stabilized yet. The merge target is not master but rather a staging branch. We will have a chance to introduce emergency breaking changes later before finally merging things to master (that would be after the implementations have been updated and some initial feedback has been gathered).

pavel-kirienko commented:
> I would prefer the "Dynamic" range to be changed to an "Application-specific" range before merging though.

Done.

> Can we also clarify in the README.md that this is currently a draft as well?

Not done -- I'm worried we might forget to remove that before merging to master. The facts that the spec v1.0 is not yet released and that this is neither master nor a release branch should be indicative enough.

I would like @thirtytwobits and @antoinealb to confirm that they accept the changes as well before merging. After that we will continue with `diagnostics`.

thirtytwobits commented on uavcan/node/430.GetInfo.0.1.uavcan:
```diff
@@ -0,0 +1,69 @@
+# The certificate of authenticity (COA) of the node, 255 bytes max, optional.
+# E.g., this field can be used for reporting RSA signatures (up to RSA-2040).
+# Leave empty if not used. Not to be changed while the node is runnning.
+uint8[<256] certificate_of_authenticity
```
I'd move this to the end. There's just something intuitive about having this massive data buffer be the last field.

pavel-kirienko replied:
Done. I have rearranged the entire message to move dynamically sized fields to the end.

thirtytwobits commented on README.md:
```diff
@@ -11,4 +10,140 @@ UAVCAN is an open lightweight protocol designed for reliable communication in ae
+----------------------------|-------------------------
+`uavcan.time`               | 62804
+`uavcan.node`               | 62805
+`uavcan.pnp`                | 62810/65533
```
I'm not a fan of the PNP name. There have been so many security-hole-riddled things with the name PNP that it's sort of tainted. Furthermore, I'm interested in the assignment of dynamic node IDs as a one-time operation which doesn't really fit into the concept of PNP.

How about `uavcan.config` instead? This is ultimately a configuration step and there may be other standard configuration functions that emerge in the future.

pavel-kirienko replied:
Eh, our node ID allocation protocol does not concern itself with security against malicious nodes either, that would be outside of the scope of the protocol. Just wash your ~hands~ nodes before connecting them to the bus.

Config is bad, sounds like something about configuration parameters. Let's look for better names please. I personally like (despite bad PR) PNP because it's to the point and you understand what it's about when you look at it, but I don't insist on that.

> I'm interested in the assignment of dynamic node IDs as a one-time operation which doesn't really fit into the concept of PNP.

It is, by the very definition of the protocol, a one-time operation, and it is guaranteed even by the current spec v0 (quote from uavcan.org):

> During dynamic allocation, the allocatee communicates its unique ID to the allocator (or allocators), which then use it to produce an appropriate allocation response. Unique ID values are kept by allocators in allocation tables -- data structures that contain mappings between unique ID and corresponding node ID values. **Allocation tables are write-only data structures that can only grow. Once a new allocatee has requested a node ID, its unique ID will be recorded into the allocation table, and all subsequent allocation requests from the same allocatee will be served with the same node ID value.**

thirtytwobits replied:
okay. pnp is fine.

thirtytwobits commented on uavcan/time/Point.1.0.uavcan:
```diff
@@ -0,0 +1,11 @@
+
+# The number of microseconds that have passed since some arbitrary moment in the past.
+# This value is to never overflow.
+truncated uint56 microseconds
```
(why is this not 64 bits?)

pavel-kirienko replied:
Added comments explaining that. In short, there are two reasons:

* Max 7 bytes per single-frame transfer in CAN 2.0
* Existential crisis

pavel-kirienko commented on uavcan/node/430.GetInfo.0.1.uavcan:
```diff
@@ -0,0 +1,69 @@
+# For example, this field can be used for reporting the short git commit hash of the current
+# software revision.
+# Set to zero if not used.
+uint32 software_vcs_revision_id
```
why though

pavel-kirienko commented on uavcan/node/430.GetInfo.0.1.uavcan:
```diff
@@ -0,0 +1,69 @@
+Version.1 software_version
+
+# Reserved for future use
+void23
```
The question is no longer applicable because I've moved stuff around, but I should answer nevertheless for the benefit of posterity: the next fields needed 1 bit for length, so 23+1 gave us 3 bytes. The preceding fields were 13 bytes long, so the resulting offset at the CRC field was 16 byte.

pavel-kirienko commented on uavcan/internet/README.md:
```diff
@@ -0,0 +1,16 @@
+
+## Application examples
+
+* Direct telemetry transmission from UAVCAN nodes with no dependency on external means of broadcasting.
```
Bad wording. What I meant is that one can send data out using a well-defined UAVCAN interface rather than relying on the autopilot or some other highly-capable node to tunnel stuff outside for them in an application-specific non-cross-vendor-compatible way. I have fixed this.

pavel-kirienko commented:
> Some thoughts about the file API (just food for thought. You're the architect here so I trust your final judgment): I'm wondering if we have an API that is weird? That is, we could have used something like TFTP over CAN for greater simplicity and standardization. On the other hand, or we could have provided something more complete like HTTP over CAN (e.g. using HTTP verbs for basic operations and RFC7233 range requests for more fine grained file access). With that in mind, is what we have here a happy median or an odd bastard?

@thirtytwobits I would say that what we have here is more of a happy bastard. I seriously dislike conventional file transfer protocols because of their heavy reliance on shared state which makes them extremely fragile; TFTP is a good example of that. Instead, we should be leaning towards client-driven protocols that ideally do not require the server to keep any context whatsoever (although this is probably not possible or not practical to obtain, it is something we should strive towards, like a motivational poster on the wall). Let's discuss that in detail in a separate PR.

thirtytwobits commented on uavcan/node/430.GetInfo.0.1.uavcan:
```diff
@@ -0,0 +1,69 @@
+# For example, this field can be used for reporting the short git commit hash of the current
+# software revision.
+# Set to zero if not used.
+uint32 software_vcs_revision_id
```
```suggestion
uint64 software_vcs_revision_id
```
for future-proofing.

pavel-kirienko replied:
For that to make sense, there should be either:

- More than 2^32 commits.
- The number of commits should be large enough for a short hash collision to be likely. Consider this: assuming that the short hash is perfect, you need ~3000 commits to reach the collision probability of 0.1% with a 32-bit hash. 3k commits is not much (for reference, there are 28k commits in the PX4 Firmware repository), but you also have the major and the minor version numbers to further narrow down the exact revision, so your real constraint is that there shouldn't be more than a couple thousand of commits sharing the same minor version number. Which seems sensible.

Do we really want 64 bits here?

antoinealb replied:
I agree with @thirtytwobits here, I would prefer 64 bits of vcs commit it, especially since since GetInfo is not going to be called very often anyway.

pavel-kirienko replied:
@antoinealb what is the rationale?

antoinealb replied:
If I computed correctly, at 28k commit, you have about 9% chance of a collision in the 32 bit git hash. Of course like you said it can be narrowed down using major and minor, but that's a hassle that is really easy to avoid. If we use 64 bits instead for example, the probability drops to about 10e-11.

pavel-kirienko replied:
> If we use 64 bits instead for example, the probability drops to about 10e-11.

Okay. That is probably an overkill, how about we used 40-bit here? For 40-bit hash:

* ~0.1% at 50k commits
* ~24% at 784k commits (that's how many there are in the Linux kernel currently)

For 48-bit hash:

* ~0.0004% at 50k commits
* ~0.1% at 784k commits

antoinealb replied:
40 or 48 are ok for me.

pavel-kirienko replied:
Done, increased to 40 bit.

thirtytwobits commented on uavcan/node/435.ExecuteCommand.0.1.uavcan:
```diff
@@ -0,0 +1,68 @@
+#
```
what do you think about calling this SystemCall?

pavel-kirienko replied:
It should be a verb, so `ExecuteSystemCall`? It is a higher-level action, not really a system call, isn't it?

thirtytwobits commented on uavcan/node/435.ExecuteCommand.0.1.uavcan:
```diff
@@ -0,0 +1,68 @@
+# A string parameter supplied to the command. The format and interpretation is command-specific.
+# The standard commands do not use this field (ignore it), excepting the following:
+#   - COMMAND_BEGIN_SOFTWARE_UPDATE
+uint8[<=200] parameter
```
I'd call this "payload" but I'm not very good at naming things, apparently.

pavel-kirienko replied:
The objective here is not to transfer some data to the server, but to enrich the command with additional data that specifies its behavior. In that sense, "payload" doesn't work well, but "argument" or "data" do.

thirtytwobits replied:
Okay. I don't feel too strongly about this. parameter is fine. My thinking with "payload" is that sometimes (perhaps most of the time) this will be a parameter but sometimes it will be data. "Payload" is more of an generic name for bits that can be either.

thirtytwobits commented on uavcan/node/435.ExecuteCommand.0.1.uavcan:
```diff
@@ -0,0 +1,68 @@
+uint8 STATUS_FAILURE        = 1     # Could not start or execution failed
+uint8 STATUS_DENIED         = 2     # The server understood the command but refused to execute it
+uint8 STATUS_BAD_COMMAND    = 3     # The requested command is not known or not supported
+uint8 STATUS_BAD_PARAMETER  = 4     # The supplied parameter cannot be used with the selected command
```
```suggestion
uint8 STATUS_BAD_PARAMETER  = 4     # The supplied parameter cannot be used with the selected command
uint8 STATUS_CONFLICT       = 5     # The operation could not be completed because of a conflicting state
```

pavel-kirienko replied:
Looks like a special case of `STATUS_DENIED`. Perhaps we should rename `DENIED` or come up with a completely different set of status codes?

thirtytwobits replied:
You added BAD_STATE which is a synonym for CONFLICT so this doesn't need to be added now.

thirtytwobits commented on uavcan/node/435.ExecuteCommand.0.1.uavcan:
```diff
@@ -0,0 +1,68 @@
+# The result of the request.
+uint8 STATUS_SUCCESS        = 0     # Started or executed successfully
+uint8 STATUS_FAILURE        = 1     # Could not start or execution failed
+uint8 STATUS_DENIED         = 2     # The server understood the command but refused to execute it
```
```suggestion
uint8 STATUS_NOT_AUTHORIZED = 2     # The server understood the operaiton but refused to execute it for the requesting node
```

i.e. Some other node might be able to execute this operation or the requesting node might be allowed after it completes some undefined authentication step. This change goes with the addition of `STATUS_CONFLICT` where a conflict would mean that the node understood but wouldn't do the operation at this time (i.e. while it is in a given state).

If practicing defense-in-depth  the implementer can choose to return `STATUS_BAD_COMMAND` to avoid exposing "open operations" to unauthenticated nodes on the bus.

pavel-kirienko replied:
I see what you meant above now. I am going to add `STATUS_NOT_AUTHORIZED` and `STATUS_BAD_STATE` (the latter is to replace obscurely named `STATUS_CONFLICT`).

thirtytwobits commented on uavcan/node/435.ExecuteCommand.0.1.uavcan:
```diff
@@ -0,0 +1,68 @@
+uint16 COMMAND_RESTART = 65535
+
+# Shut down the node; further access will not be possible until the power is turned back on.
+uint16 COMMAND_POWER_OFF = 65534
```
should this be `COMMAND_ENTER_LOW_POWER` or `COMMAND_SLEEP` instead?

pavel-kirienko replied:
I think "power off" conveys the intent better. This is mostly added to support on-board equipment that is permanently connected to the on-board power supply that should be able to somehow disable its own functionality for power conservation purposes when the vehicle is not in use. I agree that in most cases "low-power mode" or "sleep mode" would be adequate, but I don't want to impose any restrictions on how the power conservation feature should be actually implemented, hence I chose a very generic term "power off".

thirtytwobits commented on uavcan/node/435.ExecuteCommand.0.1.uavcan:
```diff
@@ -0,0 +1,68 @@
+uint8 STATUS_DENIED         = 2     # The server understood the command but refused to execute it
+uint8 STATUS_BAD_COMMAND    = 3     # The requested command is not known or not supported
+uint8 STATUS_BAD_PARAMETER  = 4     # The supplied parameter cannot be used with the selected command
+uint8 status
```
```suggestion
uint8 STATUS_INTERNAL_ERROR = 6     # The operation should have succeeded but an unexpected failure occurred.
uint8 status
```

pavel-kirienko replied:
Is it not the same thing as `STATUS_FAILURE`?

thirtytwobits replied:
It's more of an exceptional failure. I'm imagining something like:
```
function doCommandFoo()
{
    try
    {
        retval = (tryToDoIt() ? STATUS_SUCCESS : STATUS_FAILURE)
    }
    except
    {
        retval = STATUS_INTERNAL_ERROR
    }
    return retval
}
```

thirtytwobits replied:
It's more of an exceptional failure. I'm imagining something like:
```
function doCommandFoo()
{
    try
    {
        retval = (tryToDoIt() ? STATUS_SUCCESS : STATUS_FAILURE)
    }
    except
    {
        retval = STATUS_INTERNAL_ERROR
    }
    return retval
}
```

thirtytwobits replied:
It's more of an exceptional failure. I'm imagining something like:
```
function doCommandFoo()
{
    try
    {
        retval = (tryToDoIt() ? STATUS_SUCCESS : STATUS_FAILURE)
    }
    except
    {
        retval = STATUS_INTERNAL_ERROR
    }
    return retval
}
```

thirtytwobits commented on uavcan/node/62805.Heartbeat.1.0.uavcan:
```diff
@@ -0,0 +1,69 @@
+uint3 MODE_OFFLINE          = 7
+
+# Optional, vendor-specific node status code, e.g. a fault code or a status bitmask.
+truncated uint19 vendor_specific_status_code
```
`internal_status_code` perhaps? I'm okay with the static typing.

pavel-kirienko replied:
The intention of this field is to report a condensed status information in a vendor-specific way, hence the name. I don't think any of the suggested so far alternatives suit this use better.
